# NIFTY50 Market Direction Prediction

## Notebook 01 — Data Collection

### Objective

This notebook is responsible for collecting, validating, and storing all raw datasets used in the project.

No preprocessing, feature engineering, or modeling is performed here.

## Workflow

1. Download historical market data
2. Verify data quality
3. Inspect columns and date ranges
4. Save raw datasets

In [1]:
# Import libraries

import pandas as pd
import numpy as np

In [2]:
import pandas as pd
import numpy as np
import yfinance as yf

In [25]:
# Project constants

NIFTY_TICKER = "^NSEI"
DATA_FOLDER = "data/raw"

In [3]:
print(pd.__version__)

2.2.2


In [4]:
#!pip install yfinance


In [5]:
import yfinance as yf

In [6]:
# Download complete available history

nifty = yf.download(
    "^NSEI",
    period="max",
    auto_adjust=False
)

[*********************100%***********************]  1 of 1 completed


In [7]:
nifty.head()
nifty.tail()
nifty.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4621 entries, 2007-09-17 to 2026-07-21
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (Adj Close, ^NSEI)  4621 non-null   float64
 1   (Close, ^NSEI)      4621 non-null   float64
 2   (High, ^NSEI)       4621 non-null   float64
 3   (Low, ^NSEI)        4621 non-null   float64
 4   (Open, ^NSEI)       4621 non-null   float64
 5   (Volume, ^NSEI)     4621 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 252.7 KB


In [8]:
# Official Yahoo Finance ticker
NIFTY_TICKER = "^NSEI"

In [9]:
nifty = yf.download(
    NIFTY_TICKER,
    period="max",
    auto_adjust=False
)

[*********************100%***********************]  1 of 1 completed


In [10]:
nifty.head()
nifty.tail()
nifty.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4621 entries, 2007-09-17 to 2026-07-21
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (Adj Close, ^NSEI)  4621 non-null   float64
 1   (Close, ^NSEI)      4621 non-null   float64
 2   (High, ^NSEI)       4621 non-null   float64
 3   (Low, ^NSEI)        4621 non-null   float64
 4   (Open, ^NSEI)       4621 non-null   float64
 5   (Volume, ^NSEI)     4621 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 252.7 KB


In [11]:
print(nifty.columns)

MultiIndex([('Adj Close', '^NSEI'),
            (    'Close', '^NSEI'),
            (     'High', '^NSEI'),
            (      'Low', '^NSEI'),
            (     'Open', '^NSEI'),
            (   'Volume', '^NSEI')],
           names=['Price', 'Ticker'])


In [12]:
def clean_yfinance_columns(df):
    """
    Convert yfinance MultiIndex columns into standard single-level columns.
    """

    if isinstance(df.columns, pd.MultiIndex):
        df.columns = df.columns.get_level_values(0)

    return df

In [13]:
nifty = clean_yfinance_columns(nifty)

In [14]:
print(nifty.columns)

Index(['Adj Close', 'Close', 'High', 'Low', 'Open', 'Volume'], dtype='object', name='Price')


In [15]:
nifty.head()

Price,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [16]:
nifty.columns.name = None

In [17]:
nifty.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [18]:
nifty.describe()

,Adj Close,Close,High,Low,Open,Volume
count,4621.000000,4621.000000,4621.000000,4621.000000,4621.000000,4.621000e+03
mean,11304.061952,11304.061952,11369.718739,11235.797270,11310.275667,2.172881e+05
std,6639.178750,6639.178750,6662.006898,6615.518884,6641.292294,2.065222e+05
min,2524.199951,2524.199951,2585.300049,2252.750000,2553.600098,0.000000e+00
25%,5718.799805,5718.799805,5755.000000,5677.899902,5717.100098,0.000000e+00
50%,8897.549805,8897.549805,8957.549805,8849.349609,8902.750000,1.940000e+05
75%,16584.300781,16584.300781,16649.199219,16447.900391,16545.250000,3.049000e+05
max,26328.550781,26328.550781,26373.199219,26210.050781,26333.699219,1.811000e+06


In [19]:
import os

os.makedirs("data/raw", exist_ok=True)

nifty.to_csv("data/raw/nifty50_raw.csv", index=True)

In [20]:
import os

print(os.listdir("data/raw"))

['nifty50_raw.csv']


In [22]:
# Read the saved dataset

nifty_saved = pd.read_csv("data/raw/nifty50_raw.csv")

In [23]:
nifty_saved.head()

,Date,Adj Close,Close,High,Low,Open,Volume
0,2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
1,2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2,2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
3,2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
4,2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [24]:
nifty_saved.shape

(4621, 7)

# Progress Summary

## Completed

- Downloaded NIFTY50 historical data
- Verified historical coverage (2007–2026)
- Standardized yfinance column names
- Validated data quality
- Saved raw dataset
- Verified saved CSV

## Dataset Summary

- Asset: NIFTY50
- Ticker: ^NSEI
- Source: Yahoo Finance (yfinance)
- Frequency: Daily
- Rows: 4621
- Columns: Date, Open, High, Low, Close, Adj Close, Volume

## Status

✅ Data collection pipeline successfully validated.